File: impl/ts/src/crypto.ts
import { sha256 as sha256hash } from "@noble/hashes/sha256";
import nacl from "tweetnacl";

export function sha256(data: Uint8Array): Uint8Array {
  return sha256hash(data);
}

export function verifyEd25519Hash(publicKey: Uint8Array, messageHash: Uint8Array, signature: Uint8Array): void {
  if (publicKey.length !== 32 || signature.length !== 64 || messageHash.length !== 32) {
    throw new Error("invalid signature");
  }
  const ok = nacl.sign.detached.verify(messageHash, signature, publicKey);
  if (!ok) {
    throw new Error("invalid signature");
  }
}

export function signEd25519Hash(secretKey: Uint8Array, messageHash: Uint8Array): Uint8Array {
  if (secretKey.length !== 32 || messageHash.length !== 32) {
    throw new Error("invalid signature inputs");
  }
  const keyPair = nacl.sign.keyPair.fromSeed(secretKey);
  return nacl.sign.detached(messageHash, keyPair.secretKey);
}

File: impl/python/agentnet_py/crypto.py
import hashlib
from nacl.signing import VerifyKey


class CryptoError(ValueError):
    pass


def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def verify_ed25519_hash(public_key: bytes, message_hash: bytes, signature: bytes) -> None:
    if len(public_key) != 32 or len(signature) != 64 or len(message_hash) != 32:
        raise CryptoError("invalid signature")
    try:
        vk = VerifyKey(public_key)
        vk.verify(message_hash, signature)
    except Exception as exc:
        raise CryptoError("invalid signature") from exc
